src/main/javascript/crypto/src/libs/AttestedObject.ts

```diff
@@ -1,34 +1,216 @@
         this.crypto = new AttestationCrypto();
         this.pok = this.makeProof(attestationSecret, objectSecret, this.crypto);
         this.derEncodedProof = this.pok.getDerEncoding();
 
-        let vec =
-            this.attestableObject.getDerEncoding() +
+        this.fillPresignData();
+
+    }
+
+    setWebDomain(domain: string){
+        this.webDomain = domain;
+    }
+
+    fillPresignData(){
+
+        this.preSignEncoded = this.attestableObject.getDerEncoding() +
             this.att.getDerEncoding() +
             this.pok.getDerEncoding();
+
+        this.encoding = Asn1Der.encode('SEQUENCE_30', this.preSignEncoded);
+    }
+
+    fromDecodedData<T extends Attestable>(
+        attestableObject: T ,
+        att: SignedIdentifierAttestation,
+        pok: ProofOfExponentInterface ,
+        signature: string
+    ){
+        this.attestableObject = attestableObject;
+        this.att = att;
+        this.pok = pok;
+        this.signature = signature;
+
+        this.fillPresignData();
+
+        let vec = this.preSignEncoded + Asn1Der.encode('BIT_STRING', this.signature)
+
         this.encoding = Asn1Der.encode('SEQUENCE_30', vec);
+        this.userKeyPair = this.att.getUnsignedAttestation().getSubjectPublicKeyInfo();
+
+        if (!this.verify()) {
+            throw new Error("The redeem request is not valid");
+        }
+    }
+
+
+    async sign(){
+        let userData = {
+            payload: this.unsignedEncoding,
+            description: AttestedObject.Eip712UserDataDescription,
+            timestamp: new Date().getTime()
+        };
+
+        return await SignatureUtility.signEIP712WithBrowserWallet(this.webDomain, userData, AttestedObject.Eip712UserDataTypes, AttestedObject.Eip712UserDataPrimaryName );
+    }
+
+    public checkValidity(): boolean {
+        // CHECK: that it is an identifier attestation otherwise not all the checks of validity needed gets carried out
+        try {
+            let attEncoded = this.att.getUnsignedAttestation().getDerEncoding();
+            let std: IdentifierAttestation = IdentifierAttestation.fromBytes(new Uint8Array(hexStringToArray(attEncoded))) as IdentifierAttestation;
+
+            // CHECK: perform the needed checks of an identifier attestation
+            if (!std.checkValidity()) {
+                console.error("The attestation is not a valid standard attestation");
+                return false;
+            }
+        } catch (e) {
+            console.error("The attestation is invalid");
+            return false;
+        }
+
+        // CHECK: that the cheque is still valid
+        if (!this.getAttestableObject().checkValidity()) {
+            console.error("Cheque is not valid");
+            return false;
+        }
+
+        // CHECK: the Ethereum address on the attestation matches receivers signing key
+        let attestationEthereumAddress: string = this.getAtt().getUnsignedAttestation().getSubject().substring(3);
+
```

Comment:
It should be checked that this also matches the signing key for the signature checked just below here.

Reply:
Specifically what is done in the java version [here](https://github.com/TokenScript/attestation/blob/22780f5207ad2a75a59ec519eb95652d48ef3776/src/main/java/com/alphawallet/attestation/AttestedObject.java#L138)

Reply:
Seems like you have already done this